chainstate/tx-verifier/src/transaction_verifier/tokens_accounting_undo_cache/mod.rs

```diff
@@ -0,0 +1,143 @@
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use super::{error::ConnectTransactionError, CachedOperation, TransactionSource};
+use common::{chain::Transaction, primitives::Id};
+use tokens_accounting::{BlockUndoError, TxUndo};
+
+mod cached_block_undo;
+pub use cached_block_undo::CachedTokensBlockUndo;
+
+pub type CachedTokensBlockUndoOp = CachedOperation<CachedTokensBlockUndo>;
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct TokensAccountingBlockUndoCache {
+    data: BTreeMap<TransactionSource, CachedOperation<CachedTokensBlockUndo>>,
+}
+
+impl TokensAccountingBlockUndoCache {
+    pub fn new() -> Self {
+        Self {
+            data: BTreeMap::new(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(
+        data: BTreeMap<TransactionSource, CachedOperation<CachedTokensBlockUndo>>,
+    ) -> Self {
+        Self { data }
+    }
+
+    pub fn data(&self) -> &BTreeMap<TransactionSource, CachedOperation<CachedTokensBlockUndo>> {
+        &self.data
+    }
+
+    pub fn consume(self) -> BTreeMap<TransactionSource, CachedOperation<CachedTokensBlockUndo>> {
+        self.data
+    }
+
+    pub fn add_tx_undo(
+        &mut self,
+        tx_source: TransactionSource,
+        tx_id: Id<Transaction>,
+        tx_undo: TxUndo,
+    ) -> Result<(), ConnectTransactionError> {
+        match self.data.entry(tx_source) {
+            Entry::Occupied(mut entry) => match entry.get() {
+                CachedOperation::Write(undo) | CachedOperation::Read(undo) => {
+                    let mut block_undo = undo.clone();
+                    block_undo.insert_tx_undo(tx_id, tx_undo)?;
+                    entry.insert(CachedOperation::Write(block_undo.clone()));
+                }
+                CachedOperation::Erase => {
+                    let block_undo =
+                        CachedTokensBlockUndo::new(BTreeMap::from_iter([(tx_id, tx_undo)]))?;
+                    entry.insert(CachedOperation::Write(block_undo));
+                }
+            },
+            Entry::Vacant(entry) => {
+                let block_undo =
+                    CachedTokensBlockUndo::new(BTreeMap::from_iter([(tx_id, tx_undo)]))?;
+                entry.insert(CachedOperation::Write(block_undo));
+            }
+        };
+
+        Ok(())
+    }
+
+    pub fn take_tx_undo<F, E>(
+        &mut self,
+        tx_source: &TransactionSource,
+        tx_id: &Id<Transaction>,
+        fetcher_func: F,
+    ) -> Result<Option<TxUndo>, ConnectTransactionError>
+    where
+        F: Fn(TransactionSource) -> Result<Option<CachedTokensBlockUndo>, E>,
+        ConnectTransactionError: From<E>,
+    {
+        let block_undo = match self.data.entry(*tx_source) {
+            Entry::Vacant(_) => fetcher_func(*tx_source)?,
+            Entry::Occupied(entry) => match entry.get() {
+                CachedOperation::Write(undo) | CachedOperation::Read(undo) => Some(undo.clone()),
+                CachedOperation::Erase => panic!("already empty"),
```

Comment:
I understand this is impossible because it's an undo, but please explain in more detail why this is impossible in expect. Short messages like this aren't a good idea in large contexts.